src/function/scalar/system/aggregate_export.cpp

```diff
@@ -0,0 +1,304 @@
+	return make_unique<FinalizeState>(bind_data.state_size);
+}
+
+static void AggregateStateFinalize(DataChunk &input, ExpressionState &state_p, Vector &result) {
+	auto &bind_data = ExportAggregateBindData::GetFrom(state_p);
+	auto &local_state = (FinalizeState &)*((ExecuteFunctionState &)state_p).local_state;
+
+	D_ASSERT(bind_data.state_size == bind_data.aggr.state_size());
+	D_ASSERT(input.data.size() == 1);
+	D_ASSERT(input.data[0].GetType().id() == LogicalTypeId::AGGREGATE_STATE);
+	auto aligned_state_size = AlignValue(bind_data.state_size);
+
+	auto state_vec_ptr = FlatVector::GetData<data_ptr_t>(local_state.addresses);
+
+	VectorData state_data;
+	input.data[0].Orrify(input.size(), state_data);
+	for (idx_t i = 0; i < input.size(); i++) {
+		auto state_idx = state_data.sel->get_index(i);
+		auto state_entry = &((string_t *)state_data.data)[state_idx];
+		auto target_ptr = (const char *)local_state.state_buffer.get() + aligned_state_size * i;
+
+		if (state_data.validity.RowIsValid(state_idx)) {
+			D_ASSERT(state_entry->GetSize() == bind_data.state_size);
+			memcpy((void *)target_ptr, state_entry->GetDataUnsafe(), bind_data.state_size);
+		} else {
+			// create a dummy state because finalize does not understand NULLs in its input
+			// we put the NULL back in explicitly below
+			bind_data.aggr.initialize((data_ptr_t)target_ptr);
+		}
+		state_vec_ptr[i] = (data_ptr_t)target_ptr;
+	}
+
+	bind_data.aggr.finalize(local_state.addresses, bind_data.bind_data.get(), result, input.size(), 0);
+
+	for (idx_t i = 0; i < input.size(); i++) {
+		auto state_idx = state_data.sel->get_index(i);
+		if (!state_data.validity.RowIsValid(state_idx)) {
+			FlatVector::SetNull(result, i, true);
+		}
+	}
+}
+
+static void AggregateStateCombine(DataChunk &input, ExpressionState &state_p, Vector &result) {
+	auto &bind_data = ExportAggregateBindData::GetFrom(state_p);
+	auto &local_state = (CombineState &)*((ExecuteFunctionState &)state_p).local_state;
+
+	D_ASSERT(bind_data.state_size == bind_data.aggr.state_size());
+
+	D_ASSERT(input.data.size() == 2);
+	D_ASSERT(input.data[0].GetType().id() == LogicalTypeId::AGGREGATE_STATE);
+	D_ASSERT(input.data[0].GetType() == input.data[1].GetType());
+	D_ASSERT(input.data[0].GetType() == result.GetType());
+
+	VectorData state0_data, state1_data;
+	input.data[0].Orrify(input.size(), state0_data);
+	input.data[1].Orrify(input.size(), state1_data);
+
+	auto result_ptr = FlatVector::GetData<string_t>(result);
+
+	for (idx_t i = 0; i < input.size(); i++) {
+		auto state0_idx = state0_data.sel->get_index(i);
+		auto state1_idx = state1_data.sel->get_index(i);
+
+		auto &state0 = ((string_t *)state0_data.data)[state0_idx];
+		auto &state1 = ((string_t *)state1_data.data)[state1_idx];
+
+		// if both are NULL, we return NULL. If either of them is not, the result is that one
+		if (!state0_data.validity.RowIsValid(state0_idx) && !state1_data.validity.RowIsValid(state1_idx)) {
+			FlatVector::SetNull(result, i, true);
```

Comment:
Perhaps we should not return `NULL` here but the empty aggregate state. This is usually `NULL` as well, but is 0 in the case of COUNT. Not sure when the scenario would occur where both sides have a `NULL`, however.

Reply:
I'll try to look at this more next week, but I was wondering when a `NULL` state can happen? IIRC the state is created with the initialise API.

Also, what happens with multithreading? And I didn't see anything for windowed aggregates?

Reply:
@hawkfish NULLs can happen when you say do a left join with a table that has a state on the left side for example. Have not looked at windowed aggrs yet indeed.

Reply:
Ah, because the LJ just sets the mask, makes sense.  And I would argue that a `NULL` state is a state of zero elements, so the semantics sound correct in all cases.

As for windowing, I'm not sure it even makes sense: why would you compute an intermediate value there? And while segment trees might work, the generalised `window` API doesn't even have states, so it wouldn't work. I vote for disallowing it entirely.

Reply:
So lets disallow it and add a test that makes sure you can't combine `OVER` and `EXPORT STATE`